### PR TITLE
Add RISC-V base support

### PR DESCRIPTION
**Pull request by YanWQ-monad**
Add RISC-V base support as described in #777.

Based on
- #1071 
- #1087

### CODE REVIEW

**Comment by tatetian** [comment]
This is a big and important PR. I would like to invite several good hands to help me out. 

@junyang-zh Could you review this PR with a focus on the changes to `ostd/`?
@StevenJiang1110  Could you review this PR with a focus on the changes to `osdk`-related stuff?
@cchanging Could you review this PR with a focus on the changes to `kernel/`?

File: kernel/aster-nix/Cargo.toml
Comment (cchanging) [inline kernel/aster-nix/Cargo.toml:76]: I'm not sure if there are other reasons for introducing the `goldfish` crate. However, from my perspective, the main reason is that the RTC part requires different implementations based on the architecture. But why can't this differentiation be handled in the original component using `cfg`? Since the public interfaces have not been changed.

File: kernel/aster-nix/src/thread/exception.rs
Comment (cchanging) [inline kernel/aster-nix/src/thread/exception.rs:14]: Add some documents to this public trait.

**Review by cchanging — COMMENTED** [review]
Good job! I'm mostly satisfied with the changes to the `kernel/` part. I've pointed out a few minor issues here

File: ostd/Cargo.toml
Comment (YanWQ-monad) [reply ostd/Cargo.toml:62]: It seems hard, because the `default` features are applied to all targets. In other words, if `intel_tdx` is included in `default` features, other non-x86 targets would also have `intel_tdx` enabled. This is weird and doesn't make sense.

I prefer something like this but it's unfortunately unsupported yet, see rust-lang/cargo#1197.
``` toml
[target.'cfg(target_arch = "x86_64")'.features]
default = ["intel_tdx", "log_color"]

[target.'cfg(target_arch = "riscv64")'.features]
default = ["log_color"]
```

We may need to enable `intel_tdx` feature manually (i.e. `--features intel_tdx`) in TDX CI.

File: kernel/aster-nix/Cargo.toml
Comment (YanWQ-monad) [reply kernel/aster-nix/Cargo.toml:76]: I'm not sure if adding all implementations of RTC to `aster-time` would make `aster-time` inflate. Because there are different implementations on different platform on RISC-V, and `goldfish` is only one of them that is mostly used in VM.

In other words, I'm not sure if it's better to have a mono `aster-time` crate or to have differents crates for differents drivers (and extract the common code to a `rtc` prerequisite library), like `aster-cmos`, `aster-goldfish`, and etc.

Comment (cchanging) [reply kernel/aster-nix/Cargo.toml:76]: I think that implementing RTC support for different platforms in the same location is better, since it can achieve more code reusing. After all, the logic for things like `tsc` and `clocksource` is consistent. We can implement the differentiated parts in different files and provide a unified interface.

File: ostd/Cargo.toml
Comment (StevenJiang1110) [reply ostd/Cargo.toml:62]: I have discussed with the TDX developer @Hsy-Intel , and he will submit a PR to place all code under `#[cfg(feature = "intel_tdx")]` to under `#[cfg(all(target_arch = "x86_64",feature = "intel_tdx" )]`, so it will be possible to set `intel_tdx` as default feature after that change?

Comment (YanWQ-monad) [reply ostd/Cargo.toml:62]: Technically, it can. But from my point of view, semantically, enabling `intel_tdx` feature on non-x86 architecture doesn't make any sense, although the actual code is not enabled.

Comment (StevenJiang1110) [reply ostd/Cargo.toml:62]: You are correct, `intel_tdx` is not compatible with riscv. Therefore, it would be best for the feature to be independent of architecture. As a result, #1111 has renamed the feature as `cvm_guest` (confidential virtual machine). Since riscv can also support its cvm, this feature may be suitable for all targets. Hopefully, this name change will help you feel comfortable in keeping it as the default feature.

File: ostd/src/arch/riscv/timer/jiffies.rs
Comment (tatetian) [inline ostd/src/arch/riscv/timer/jiffies.rs:15]: I can't see why this definition of `Jiffies` should be arch-dependent.

File: ostd/src/arch/riscv/timer/mod.rs
Comment (tatetian) [inline ostd/src/arch/riscv/timer/mod.rs:50]: I don't know about goldfish. So I did a quick search. Below is the device tree information of goldfish RTC. Since you said this driver is for golden, you should check the `compatibility` field to confirm that the driver is indeed a goldfish RTC device.

```
    rtc@101000 {
        interrupts = <0x0b>;
        interrupt-parent = <0x03>;
        reg = <0x00 0x101000 0x00 0x1000>;
        compatible = "google,goldfish-rtc";
    };
```

File: kernel/aster-nix/Cargo.toml
Comment (tatetian) [reply kernel/aster-nix/Cargo.toml:76]: I think both approaches are viable. The question is if the RTC drivers on x86 and on RISC-V can share a lot of code. If so, they should be merged into one crate. If not, they can be two parallel crates. Either way, you need to make the crate name more self-describing.

For the first approach, there will be only one crate, which is `time`. Plain and simple.

For the second approach, there should be two crates, whose names indicate their purposes. Two names that come to my mind are: `time-x86` and `time-goldfish`.

Since the data structures in `aster-goldfish` resemble those in `time`, I guess that it is more preferable to merge `aster-goldfish` into the `time` crate.

**Comment by tatetian** [comment]
This PR has been modified greatly since last review. I would like to initiate another round of review with help from these good fellows:
@junyang-zh Could you review this PR with a focus on the changes to ostd/?
@StevenJiang1110 Could you review this PR with a focus on the changes to osdk-related stuff?
@cchanging Could you review this PR with a focus on the changes to kernel/?

File: kernel/src/process/program_loader/elf/elf_file.rs
Comment (cchanging) [inline kernel/src/process/program_loader/elf/elf_file.rs:212]: -> Executable file/program

File: kernel/src/util/random.rs
Comment (cchanging) [inline kernel/src/util/random.rs:47]: Can we let the `if-else` returns `seed` and init the RNG out of the branches? This reduces code redundancy and improves readability.

File: kernel/src/util/random.rs
Comment (cchanging) [inline kernel/src/util/random.rs:42]: SeedableRng doesn't seem to be used.

File: kernel/src/arch/riscv/cpu.rs
Comment (cchanging) [inline kernel/src/arch/riscv/cpu.rs:46]: I'm not sure, but now we already have the `GeneralRegs` in ostd, we may not need this `GpRegs` any more.

**Review by cchanging — COMMENTED** [review]
Most of the changes look good, I only have a few minor suggestions.

File: ostd/src/arch/riscv/mm/mod.rs
Comment (StevenJiang1110) [inline ostd/src/arch/riscv/mm/mod.rs:89]: Use `# Safety`.

File: ostd/src/mm/heap_allocator/mod.rs
Comment (StevenJiang1110) [inline ostd/src/mm/heap_allocator/mod.rs:125]: Uncomment these lines or remove these lines.

**Review by StevenJiang1110 — COMMENTED** [review]
OSDK part looks good to me. But I'm not sure whether ktest is supported in this PR, or will be delayed to furthur PRs?

File: ostd/src/mm/heap_allocator/mod.rs
Comment (YanWQ-monad) [reply ostd/src/mm/heap_allocator/mod.rs:125]: Oh, these changed lines seems to be mis-committed, because the heap allocator be in deadlock when rescue with these `debug!` statements enabled. I'll revert them since they are not strongly related to RISC-V.

File: ostd/src/arch/riscv/device/io_port.rs
Comment (tatetian) [inline ostd/src/arch/riscv/device/io_port.rs:19]: A dummy `IoPort` type is implemented because (1) RISC-V supports and needs port-based I/O, or (2) our kernel's CPU architecture-independent code needs the `IoPort` type to compile?

File: ostd/src/arch/riscv/iommu/mod.rs
Comment (tatetian) [inline ostd/src/arch/riscv/iommu/mod.rs:27]: The framekernel architecture requires IOMMU or something equivalent to prevent malicious or mistaken DMA that corrupts typed memory. So I googled "RISC-V IOMMU"; there is indeed a RISC-V IOMMU specification. We should and will implement IOMMU in the future. So I suggest adding a comment to state this intention.

```rust
pub(crate) fn init() -> Result<(), IommuError> {
    // TODO: We will support IOMMU on RISC-V
    Err(IommuError::NoIommu)
}
```

File: ostd/src/arch/riscv/mod.rs
Comment (tatetian) [inline ostd/src/arch/riscv/mod.rs:22]: This method should be given an arch-independent name such as `init_cvm_guest` since there won't be TDX on RISC-V.

File: ostd/src/arch/riscv/mod.rs
Comment (tatetian) [inline ostd/src/arch/riscv/mod.rs:28]: Incomplete TODO sentence

File: ostd/src/arch/riscv/mm/mod.rs
Comment (junyang-zh) [inline ostd/src/arch/riscv/mm/mod.rs:152]: The implementation in x86 have introduced a macro to make life easier:

https://github.com/asterinas/asterinas/blob/0a36760f7a42b7fe27735e29f971423f6ec4b1ca/ostd/src/arch/x86/mm/mod.rs#L150-L155

https://github.com/asterinas/asterinas/blob/0a36760f7a42b7fe27735e29f971423f6ec4b1ca/ostd/src/arch/x86/mm/mod.rs#L187-L191

File: ostd/src/arch/riscv/qemu.rs
Comment (junyang-zh) [inline ostd/src/arch/riscv/qemu.rs:18]: Do QEMU-riscv support ISA-DEBUG-EXIT device? I see the implementation does somewhat uses SBI calls to shut QEMU down. So perhaps the documentation doesn't match the implementation.

File: ostd/src/arch/riscv/trap/mod.rs
Comment (junyang-zh) [inline ostd/src/arch/riscv/trap/mod.rs:42]: I believe that IS_KERNEL_INTERRUPTED should only be set in the `Trap::Interrupt(_) => ` branch. If the calling code is in a CPU exception handler, `is_kernel_interrupted()` just returns false.

File: ostd/src/mm/heap_allocator/mod.rs
Comment (junyang-zh) [reply ostd/src/mm/heap_allocator/mod.rs:125]: I think that it has been fixed by #1368

**Review by junyang-zh — COMMENTED** [review]
Nice work!

File: kernel/src/util/random.rs
Comment (YanWQ-monad) [reply kernel/src/util/random.rs:42]: [`StdRng::from_seed`](https://docs.rs/rand/latest/rand/rngs/struct.StdRng.html#method.from_seed) is from trait `SeedableRng`.

File: kernel/src/process/program_loader/elf/elf_file.rs
Comment (YanWQ-monad) [reply kernel/src/process/program_loader/elf/elf_file.rs:212]: "Executable" can also be noun? I'm not sure if it's necessary to add "file"/"program".

File: ostd/src/arch/riscv/device/io_port.rs
Comment (tatetian) [reply ostd/src/arch/riscv/device/io_port.rs:19]: Let's leave this issue to future PRs. We can't make everything right in a single PR.

**Review by tatetian — APPROVED** [review]
LGTM. Thanks for the contribution!

I will wait for all the other three reviewers's approvals to merge this PR.

**Review by StevenJiang1110 — APPROVED** [review]

**Review by junyang-zh — APPROVED** [review]
I am OK with it!

**Comment by junyang-zh** [comment]
You should somewhat press the typo checker about it's complain on `sie`, which is a register name.

**Review by cchanging — APPROVED** [review]

**Pull request merged**
